script/domVisitor.ts

```diff
@@ -0,0 +1,124 @@
+// tslint:disable:no-console
+import * as dom from '@openstax/types/lib.dom';
+import { dirname } from 'path';
+import puppeteer from 'puppeteer';
+
+const rootUrl = `http://localhost:${process.env.PORT || '8000'}`;
+const devTools = false;
+const onlyOneBook = process.argv[2];
+
+function browserFindMatches(): string[] {
+  // Note: This executes in the browser context
+  if (!document) {
+    throw new Error(`BUG: Could not find document in the browser.`);
+  }
+
+  function nearestId(el: dom.Element): string {
+    const id = el.getAttribute('id');
+    if (id) {
+      return id;
+    }
+    if (el.parentElement) {
+      return nearestId(el.parentElement);
+    }
+    throw new Error('BUG: Could not find an ancestor with an id');
+  }
+
+  // Remove all the MJX_Assistive_MathML elements because they
+  // can be wider even though they are not visible
+  Array.from(document.querySelectorAll('.MJX_Assistive_MathML')).forEach((el) => el.remove());
+
+  const wideIds = new Set<string>();
+  const root = document.querySelector('#main-content > div');
+  if (!root) {
+    throw new Error(`BUG: Could not find content root`);
+  }
+  const rootRight = root.getBoundingClientRect().right;
+  Array.from(root.querySelectorAll('*')).forEach((c) => {
+    if (c.getBoundingClientRect().right > rootRight) {
+      wideIds.add(nearestId(c));
+    }
+  });
+  return Array.from(wideIds);
+}
+
+async function run() {
+  const browser = await puppeteer.launch({
+    devtools: devTools,
+  });
+  const page = await browser.newPage();
+  page.setDefaultNavigationTimeout(60 * 1000);
+
+  const books = await findBooks(page);
+
+  for (const book of books) {
+    const bookPages = await findBookPages(page, book.href);
+
+    let index = 0;
+    for (const bookPageUrl of bookPages) {
+      const pageUrl = `${dirname(book.href)}/${bookPageUrl}`;
+      console.log(`Checking ${index}/${bookPages.length}`);
+
+      await page.goto(`${rootUrl}/${pageUrl}`);
+      await page.waitForSelector('body[data-rex-loaded="true"]');
+
+      const matches = await page.evaluate(browserFindMatches);
+      reportMatches(pageUrl, matches);
+      index++;
+    }
+
+  }
+
+  await browser.close();
+
+  if (books.length === 0) {
+    console.error(`Could not find a matching book. ${onlyOneBook ? 'Check that the slug name is correct' : ''}`);
+    process.exit(1);
+  }
+}
+
+run().then(null, (err) => {
+  console.error(err);
+  process.exit(1);
+});
+
+async function findBooks(page: puppeteer.Page) {
+  await page.goto(rootUrl);
+  await page.waitForSelector('a[data-slug]');
```

Comment:
tying this to the developer page worries me a little. if the content is prerendered you could load the release.json, or if you want to couple it to running locally you could just load the books configuration in this script

Reply:
I wanted it to work on both prerendered (faster) and dev. Is there a way to get the list of books regardless of which mode the server is running in?

Reply:
if you're running locally you can use the BOOKS environment variable, if you're running remotely you can access the release.json file

Reply:
hmm, it seems like both `release.json` and `config.development.js` contain the book UUID rather than the openstax book slug. The only place I've found that contains the book slug is in the app in the development page. Is there another way to get the book slug?

Reply:
it would be at `http://localhost:8000/rex/release.json` but it is only for pre-rendered deployments and only has the id. other scripts (like prerender) fetch the slug from the id when they run